src/nvim/getchar.c

```diff
@@ -1644,6 +1644,342 @@ static int handle_int(const int advance)
 
   return c;
 }
+
+// Returns 0 to continue, -1 to carry on in the loop, otherwise, returns the
+// character that should be used.
+// Keeps track of the mapdepth in the variable pointed to by "mapdepthp".
+static int look_in_typebuf(int *mapdepthp, int *keylenp,
+    int *mp_match_lenp,
+    const int timedout, const int advance, const int local_State)
```

Comment:
The main point of extracting this is to mark clearly which local variables are used, and which variables are modified.
I find it much easier to reason about a function if I can know for certain that in a block of code (for example) variable `timedout` is not modified, variable `mode_deleted` is not used etc. and the main problem I have with the current `vgetorpeek()` function is that it has too much local state shared over a lot of code that doesn't need it.

As a side-benefit it does make it much easier for someone new to the code to see what's going on in `vgetorpeek()`: look in the typebuffer, if nothing there, get keys from user, ad infinitum.

The coupling to the current state of the world is surprisingly small (mainly through `typebuf`), and moreover the coupling to the action of the block that I turn into `get_key_from_user()` (both through the global state, and directly) is small too.
Other than working with `typebuf`, the only global state that the two both work with is via calls to update the ui that are put after setting `'paste'` in `look_in_typebuf()` and after the `ESC` special case and the addition of pretty characters when typing an `INSERT|CMDLINE` mode map in `get_key_from_user()`.

I agree that there are downsides in putting the code elsewhere, in a perfect world I would use a hypothetical language construct where I can just say "this block can read local variables X, and Y, and can write variables A, B".
In this case, I feel the benefits of marking what variables are used and what aren't, coupled with the ease-of-reading outweigh the downsides, but that's only my opinion.